Clean up QueryParseContext and don't hold it inside QueryRewrite/ShardContext

cbuescher opened the pull request:
This change cleans up a few things in QueryParseContext and QueryShardContext that make it hard to reason about the state of these context objects and are thus error prone and should be simplified.

Currently the parser that used in QueryParseContext can be set and reset any time from the outside, which makes reasoning about it hard.
This change makes the parser a mandatory constructor argument removes ability to later set a differen ParseFieldMatcher. If none is provided at construction time, the one set inside the parser is used. If a ParseFieldMatcher is specified at construction time, it is implicitely set on the parser that is beeing used. The ParseFieldMatcher is only kept inside the parser.

Also currently the QueryShardContext historically holds an inner QueryParseContext (in the super class QueryRewriteContext), that is mainly used to hold the parser and parseFieldMatcher. For that reason, the parser can also be reset, which leads to the same problems as above. This change removes the QueryParseContext from QueryRewriteContext and removes the ability to reset or retrieve it from the QueryShardContext. Instead, `QueryRewriteContext#newParseContext(parser)` can be used to create new parse contexts with the given parser from a shard context when needed.

cbuescher committed:
Clean up QueryParseContext and don't hold it inside QueryRewrite/ShardContext

This change cleans up a few things in QueryParseContext and QueryShardContext
that make it hard to reason about the state of these context objects and are
thus error prone and should be simplified.

Currently the parser that used in QueryParseContext can be set and reset any
time from the outside, which makes reasoning about it hard. This change makes
the parser a mandatory constructor argument removes ability to later set a
different ParseFieldMatcher. If none is provided at construction time, the
one set inside the parser is used. If a ParseFieldMatcher is specified at
construction time, it is implicitely set on the parser that is beeing used.
The ParseFieldMatcher is only kept inside the parser.

Also currently the QueryShardContext historically holds an inner QueryParseContext
(in the super class QueryRewriteContext), that is mainly used to hold the parser
and parseFieldMatcher. For that reason, the parser can also be reset, which leads
to the same problems as above. This change removes the QueryParseContext from
QueryRewriteContext and removes the ability to reset or retrieve it from the
QueryShardContext. Instead, `QueryRewriteContext#newParseContext(parser)` can be
used to create new parse contexts with the given parser from a shard context
when needed.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:
```diff
@@ -42,7 +43,6 @@ public QueryRewriteContext(IndexSettings indexSettings, MapperService mapperServ
         this.scriptService = scriptService;
         this.indexSettings = indexSettings;
         this.indicesQueriesRegistry = indicesQueriesRegistry;
-        this.parseContext = new QueryParseContext(indicesQueriesRegistry);
```
so we had this instance member but it was unused? :)

javanna replied:
oh I see better now that I opened up the code....what a mess :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -26,42 +26,33 @@
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
```
does it still make sense to use a parse field matcher that is not the one set to the parser at this point ? maybe too big of a change for now?

cbuescher replied:
I added this mainly because in tests we sometimes use ParseFields.EMPTY, sometimes ParseFields.STRICT. I think it would generally be much better to only keep the ParseFieldMatcher in one place, e.g. the parser itself, so I will look if I can get rid of this.

cbuescher replied:
Comming back to this: the problem with the parseFieldMatcher stored inside AbstractXContentParser is that (currently on maste) we don't set it anywhere except in ObjectParserTests and when the parser is reset() on the QueryParseContext (where its broken btw.). I agree that it would be best if there was only one ParseFieldMatcher (either the one set on the parser, or the one in the QueryParseContext). 

Having PFM only in the parser is hard because:
* we'd ideally make parseFieldMatcher a ctor arg for all XContentParsers. This however is a huge change (tons of ways to create XContentParser, all would need to be changes), I'm not even sure if this would work because we would need the IndexSettings (where the flag for strict parsing comes from) is held
* if we keep the current get/setParseFieldMatcher on XContentParser it is hard to check that we alway set it correctly

On the other hand, keeping PFM only in the QueryShardContext requires us removing it from ObjectParser. I ran into many problems trying to do this (to me it seems the setters, getters where just introduced in 5f258b to make PFM play nicely with ObjectParser). 

So for now I'm a little bit stuck here. I'm leaning toward trying to removing the PFM from the parser and only store it in the QueryParseContext, but I'm struggling to remove it there because ObjectParser depends and so far I didn't find a way to get around the getter/setter on the parser.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -26,42 +26,33 @@
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
+        this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
+        this.parser = Objects.requireNonNull(parser, "parser cannot be null");
+        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```
refers to the other comment above, doesn't feel right to mutate one of the arguments here based on the second argument. Can we require that the parse field matcher is properly set to the parser all the time instead?

javanna replied:
fixing this would allow us to get rid of setParseFieldMatcher in XContentParser and make the matcher final in AbstractXContentParser. I think this should be the direction moving forward, what do you think?

cbuescher replied:
Again I agree, just couldn't decide where the best place for the one "authoritative" source of the parse field matcher that is used should be. I'll see if we can have it only in the parser.

cbuescher replied:
So just from taking a brief look at this, this going to be a big PITA: making the matcher final in AbstractXContentParser would mean we need it as ctor argument, which means we need the index setting everwhere we create a parser, which is basically in a gazillion of places. Actually the parseFieldMatcher inside AbstractXContentParser was added just recently, I think mainly to be available in ObjectParser and RestReindexAction, where we up to now dont have a ParseContext. The problem is that currently the ParseFieldMatcher in the context and the one used in the parser might diverge.

javanna replied:
ok, I haven't followed all recent changes, I just realized that there are ParseFieldMatcher instances in different places. I also wonder now if it should be in the parser itself, like should it be in the parser or in the parse context? We should pick one, it cannot be both ideally :)

cbuescher replied:
As I said, ideally I'd say its the parser, but thats a huge change. Also one might argue that the parsers job is to keep state for parsing, but the matcher is something the client (e.g. the builders fromXContent) use, so there is just the question of how to transport it there, so that would be the context. I'm just looking into how difficult it would be to remove the parseField from the parser again, this duplication is a source of errors.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
```
I hadn't noticed this before, but isn't there a circular dependency between IndexSettings and ParseFieldMatcher. I think I would remove the instance within IndexSettings if possible. Doesn't seem hard.

cbuescher replied:
So you suggest removing the parseFieldMatcher from IndexSettings and and instead hold an instance insice QueryShardContext instead? As far as I can see we need IndexSettings in QueryShardContext for the super ctor anyway, so my idea was to have it only in one place. Keeping the "real" instance in IndexSettings has the advantage that it only has to be creates once, while when moving it to QueryShardContext, we will have to create it in the ctor every time, but maybe I missunderstand what you mean.

javanna replied:
forget about it, ParseFieldMatcher depends on Settings not IndexSettings, no circular dep, no problem! Leave it as is, your explanation makes perfect sense.

javanna replied:
I wish we were creating less instances of ParseFieldMatcher actually, but that's another problem :)

cbuescher replied:
++ to that. In the tests its good though that some methods take the matcher as an argument, so we can test thing in strict and non-strict mode without having pass IndexSettings around.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
+        return indexSettings.getParseFieldMatcher();
     }
 
     public void reset() {
```
do you think it would be possible to get rid of this reset too and make sure that each operation that needs a reset actually uses a new instance of the context? Seems much cleaner.

cbuescher replied:
I'll look into it, but maybe to keep this change from growing too large I would tackle that removal as a follow up.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -343,38 +331,35 @@ public ParsedQuery parse(XContentParser parser) {
      */
     @Nullable
     public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
```
Also, do you think it's feasible now to take all the parsing stuff out of this context? maybe the few components left that need this should rather take in as argument both shard context and parse context and use them properly instead of relying on shard context to parse through parse context?

cbuescher replied:
I agree, unfortunately these are used all over the place. I will have another look but again maybe that is better suited for a follow up PR.

javanna replied:
++

javanna commented:
I like these changes a lot! I left some comments, mainly ideas to explore I think, but they don't need to be all implemented as part of this PR. Up to you which ones can be done here and which ones are bigger and deserve their own followup PR.

One more thing, can we make the ParseFieldMatcher constructor that takes a boolean argument private?

cbuescher commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java:
```diff
@@ -190,22 +188,18 @@ static Query toQuery(QueryShardContext context, boolean mapUnmappedFieldsAsStrin
         // as an analyzed string.
         context.setAllowUnmappedFields(false);
         context.setMapUnmappedFieldAsString(mapUnmappedFieldsAsString);
-        context.parseFieldMatcher(context.getIndexSettings().getParseFieldMatcher());
```
The shard context parseFieldMatcher IMHO should always be backed by `indexSettings.getParseFieldMatcher()`

cbuescher commented:
@javanna I rebased this PR on current master and added a commit that now keeps the parseFieldMatcher only in the parser and not storing it in the QueryParseContext. The `parseFieldMatch()` getter of the context just forwards to the parser. This however makes is necessary that we alway set the parseFieldMatcher in the parsers instead of now where we set it on the context. In the future we should try to make ParseFieldMatcher immutable on parsers and make it a mandatory constructor argument of AbstractXContentParser. But this is a big change and out of the scope of this PR.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -343,38 +331,38 @@ public ParsedQuery parse(XContentParser parser) {
-    public Query parseInnerQuery() throws IOException {
-        return toQuery(this.parseContext().parseInnerQueryBuilder(), this);
+    public Query parseInnerQuery(XContentParser parser) throws IOException {
+        parser.setParseFieldMatcher(this.parseFieldMatcher());
```
it is odd that we do the opposite here, as the context holds the parse field matcher and sets it to the parser. But I do see why. The parse field matcher is used for more than just parsing xcontent. I think it's ok then.

javanna commented on core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java:
```diff
@@ -162,8 +162,8 @@ protected boolean doEquals(WrapperQueryBuilder other) {
         try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
-            QueryParseContext parseContext = context.newParseContext();
-            parseContext.reset(qSourceParser);
+            qSourceParser.setParseFieldMatcher(context.indexSettings.getParseFieldMatcher());
```
do we want to expose getParseFieldMatcher in QueryRewriteContext that takes it from indexSettings? then we would not need it in QueryShardContext as it's a subclass?

cbuescher replied:
good point, I think thats a good idea. Only one is called parseFieldMatcher, the other getParseFieldMatcher, I guess we are leaning to getParseFieldMatcher then.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -37,10 +37,9 @@
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser) {
         this.indicesQueriesRegistry = Objects.requireNonNull(registry, "indices queries registry cannot be null");
         this.parser = Objects.requireNonNull(parser, "parser cannot be null");
-        this.parser.setParseFieldMatcher(Objects.requireNonNull(parseFieldMatcher, "parse field matcher cannot be null"));
```
By moving this out we need to make sure the setter is called in many places where we previously implicitly set it via `context.parseFieldMatcher(pfm)`. I only checked the places where previously this method was calles somehow. 
I'm on the fence if maybe we should instead keep it here, that way at least we know there is one location where the parseFieldMatcher is set.

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorParsers.java:
```diff
@@ -131,6 +131,7 @@ public AggregatorParsers(ParseFieldRegistry<Aggregator.Parser> aggregationParser
                     if (token == XContentParser.Token.VALUE_STRING || token == XContentParser.Token.VALUE_EMBEDDED_OBJECT) {
                         byte[] source = parser.binaryValue();
                         binaryParser = XContentFactory.xContent(source).createParser(source);
+                        binaryParser.setParseFieldMatcher(parseContext.parseFieldMatcher());
```
seeing this I am wondering if this getter makes it confusing, shall we consider dropping it in favour of doing parseContext.parser().parseFieldMatcher() ? Or maybe it's not a big deal?

cbuescher replied:
I think its okay, maybe we can add a comment to the getter that makes clear its a shortcut to the matcher in the parser.

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -713,19 +713,14 @@ protected QB copyQuery(QB query) throws IOException {
-        queryCreationContext.reset();
-        queryCreationContext.parseFieldMatcher(ParseFieldMatcher.STRICT);
-        return queryCreationContext;
+        return new QueryShardContext(queryShardContext);
```
are we losing the STRICT bit here? it's important that we use STRICT here, so we make sure that we never output deprecated stuff ourselves. and we test deprecations separately.

cbuescher replied:
This is a shard context, so these parse field matcher should always be backed by indexSettings. If we want to default to STRICT here we should change that in the setup of the test I think.

javanna replied:
this is fixed by adding the strict setting above in the init method right?

cbuescher replied:
yes

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -713,19 +713,14 @@ protected QB copyQuery(QB query) throws IOException {
-        queryParseContext.reset(null);
-        queryParseContext.parseFieldMatcher(ParseFieldMatcher.STRICT);
+    protected static QueryParseContext createParseContext(XContentParser parser) {
+        QueryParseContext queryParseContext = new QueryParseContext(indicesQueriesRegistry, parser);
```
same here, might be that we are good, but let's make sure we don't lose the STRICT one

cbuescher commented:
@javanna so I think after opening #17756 (and after that gets some traction) I'm going to rework this and remove the last commit.

cbuescher committed:
Clean up creation of shard context in AbstractQueryTestCase

The queryShardContext we create during setup was sometimes
accessed directly, sometimes by making a copy through
the createShardContext() helper. This should be the default.
Also making sure that strict parsing is switched on via
IndexSettings in the test testup.

cbuescher committed:
Pass down parse context instead of just parser in completion context suggester.

cbuescher commented:
@javanna after merging #17756 I updated this and removed some more usages of the STRICT constant from core. I think this is much cleaner now.

javanna commented on core/src/main/java/org/elasticsearch/action/percolate/TransportPercolateAction.java:
```diff
@@ -196,9 +196,8 @@ public static SearchRequest createSearchRequest(PercolateRequest percolateReques
-            QueryParseContext queryParseContext = new QueryParseContext(queryRegistry);
-            queryParseContext.reset(XContentHelper.createParser(querySource));
-            queryParseContext.parseFieldMatcher(parseFieldMatcher);
+            QueryParseContext queryParseContext = new QueryParseContext(queryRegistry, XContentHelper.createParser(querySource),
```
just noticed that below we have try finally around the parser creation but here we don't have it. Should we have that here too? who closes the parser after all?

cbuescher replied:
Sure, I see parse creation a lot as try-with-resouces without a finally block (unless we need to reset something which here I think we don't), so I will do that.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -115,18 +114,12 @@ public QueryShardContext(QueryShardContext source) {
 
     public void reset() {
```
please please can we look into removing this as well later? :)

javanna replied:
I see that I had already asked this before, ignore me, you already said we'll do it in another PR :)

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -158,11 +157,6 @@
 
-
-    protected static QueryShardContext queryShardContext() {
-        return queryShardContext;
```
I am now wondering if we still need this queryShardContext instance member if we (almost) always create a new one.

cbuescher replied:
Its good to have it as a template for the createShardContext() method, we also need the instance member because we use it in lambdas for the mock mapping service etc...

javanna replied:
it gets a bit weird when we use it to set stuff to the SearchContext although we create another context for those operations. That seems confusing... plus if it was not for templating, it would be only needed in init() which could have it a local variable. Maybe rename it at least so it is clearer why it's there? and fix some of its usages?

cbuescher replied:
okay, so I understand you correctly: should I revert 4e77adf38 instead and we rather use the instance member as much as possible and try to remove the `createShardContext()` that makes a copy instead?

javanna replied:
no  I like creating a new instance, seems to reflect what we do everywhere at this point. But we should clean things up a little bit more, to make sure that the set search context is consistent at least.

cbuescher replied:
Okay, I think I get it. I went ahead and removed the queryShardContext instance entirely. I had to introduce a few more static fields for all the services we need in init(), but I think it is better this way. Also pulled up some mocking on to of the SearchContext that was done in the Parent/Child/Nested tests but that can be used everywhere I think.

javanna replied:
thanks a lot, looks great!

javanna commented on core/src/test/java/org/elasticsearch/search/searchafter/SearchAfterBuilderTests.java:
```diff
@@ -224,7 +221,7 @@ public void testFromXContent() throws Exception {
             builder.endObject();
             XContentParser parser = XContentHelper.createParser(builder.bytes());
-            context.reset(parser);
+            QueryParseContext context = new QueryParseContext(indicesQueriesRegistry, parser, ParseFieldMatcher.STRICT);
```
thanks, it is better to use strict than the default one here!

javanna commented on core/src/test/java/org/elasticsearch/search/sort/SortParserTests.java:
```diff
@@ -1,165 +0,0 @@
-
-import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
-
-public class SortParserTests extends ESSingleNodeTestCase {
```
oh nice we had tests for a class that didn't exist anymore :) thanks for fixing

javanna commented:
this looks much cleaner indeed! left a few minors but LGTM!

cbuescher committed:
Wrap parser creation in try-with-resources

cbuescher committed:
Cleaning up query shard context creation in AbstractQueryTestCase

cbuescher commented:
@javanna thanks for the last round of comments, I adressed them in the last two commits.

javanna commented:
LGTM
